WooCommerce/Classes/ViewRelated/Orders/Payment Methods/PaymentMethodsViewModel.swift

```diff
@@ -375,28 +389,49 @@ private extension PaymentMethodsViewModel {
     /// Observes the store CPP state and update publish variables accordingly.
     ///
     func bindStoreCPPState() {
-        ordersResultController.onDidChangeContent = updateCardPaymentVisibility
+        ordersResultController.onDidChangeContent = { [weak self] in
+            self?.updateCardPaymentVisibility()
+        }
         try? ordersResultController.performFetch()
     }
 
     func updateCardPaymentVisibility() {
         guard cardPresentPaymentsConfiguration.isSupportedCountry else {
             showPayWithCardRow = false
+            showTapToPayRow = false
 
             return
         }
 
+        localMobileReaderSupported { [weak self] tapToPaySupportedByDevice in
+            let tapToPaySupportedByStore = self?.cardPresentPaymentsConfiguration.supportedReaders.contains(.appleBuiltIn) ?? false
+            self?.orderIsEligibleForCardPresentPayment { [weak self] orderIsEligible in
+                self?.showPayWithCardRow = orderIsEligible
+                self?.showTapToPayRow = orderIsEligible && tapToPaySupportedByDevice && tapToPaySupportedByStore
+            }
+        }
+    }
+
+    private func localMobileReaderSupported(onCompletion: @escaping ((Bool) -> Void)) {
+        let action = CardPresentPaymentAction.checkDeviceSupport(siteID: siteID,
+                                                                 cardReaderType: .appleBuiltIn,
+                                                                 discoveryMethod: .localMobile,
+                                                                 onCompletion: onCompletion)
+        stores.dispatch(action)
+    }
+
+    private func orderIsEligibleForCardPresentPayment(onCompletion: @escaping (Bool) -> Void) {
         let action = OrderCardPresentPaymentEligibilityAction
             .orderIsEligibleForCardPresentPayment(orderID: orderID,
                                                   siteID: siteID,
-                                                  cardPresentPaymentsConfiguration: cardPresentPaymentsConfiguration) { [weak self] result in
-            switch result {
-            case .success(let eligible):
-                self?.showPayWithCardRow = eligible
-            case .failure(_):
-                self?.showPayWithCardRow = false
+                                                  cardPresentPaymentsConfiguration: cardPresentPaymentsConfiguration) { result in
+                switch result {
+                case .success(let eligibility):
+                    onCompletion(eligibility)
+                case .failure:
+                    onCompletion(false)
```

Comment:
Question: Just confirming, declaring `eligibility` here and using its value in `localMobileReaderSupported` is not a problem, correct? We expect that this could be both `true` or `false` on success, the actual value would depend on the `CardPresentPaymentConfiguration` details (country, payment method, currency) but is still a successful operation. I'm not quite seeing if should be like this instead:

```suggestion
                switch result {
                case .success():
                    onCompletion(false)
                case .failure:
                    onCompletion(false)
```

If any of the components of `CardPresentPaymentConfiguration` makes it non-eligible.

Reply:
Thanks! Perfectly clear! I looked too much into it since the tests at https://github.com/woocommerce/woocommerce-ios/pull/9242#discussion_r1145993970 so I wanted to be sure I wasn't missing something 😅